### PR TITLE
encrypt chat local storage with device key CORE-3946

### PR DESCRIPTION
**Pull request by mmaxim**
@maxtaco @patrickxb r?

This uses the device key to encrypt chat messages in local storage.

### CODE REVIEW

**Comment by maxtaco** [comment]
A few high-level thoughts on the encryption approach: it's probably too heavyweight in terms of computation and storage.  Computation because you need to compute a DH per message, and storage because you're paying to store the random nonce and the random ephemeral key, which is probably overkill here.

Here's an alternate approach, and paging @oconnor663 for his opinions:
(1) make a new method that calls DH precompute on the secret device key and its public counterpart
(2) call secretBox using that 32-byte key output from step (1), and the nonce is just 24 bytes of randomness (as before). We can try to skimp here if we think that 24 bytes is too expensive and just use the DB key as the nonce (since that's guaranteed not to be repeated). But then there's a slight chance of a collision with a slightly different usage.

**Comment by maxtaco** [comment]
Actually, alternatively, you can treat the device secret key as just 32-bytes, and then compute HMAC(secretKey, "Keybase-Chat-Local-Storage-1"), and use that as the secret key to SecretBox.  Then we'd solve the versioning issues that we already incurred with chat messages that @oconnor663 has already pointed out.  We'd need one more field in the `boxedLocalMessage` struct which is just `Version`

**Comment by mmaxim** [comment]
Ack yeah sorry for the newbie attempt, I'll try and get option 2 ready this morning.

**Comment by maxtaco** [comment]
Here's a proposal:

 * New typedef kind of like `SignaturePrefix`:

```
type EncryptionReason string
```

 * Change the `GenericKey` interface to include a new method that will noop/error for all keys by NaclDH

```
	SecretSymmetricKey(reason EncryptionReason) ([]byte, error)
```
 
  * For NaclDH, call:

```
        mac := hmac.New(sha256.New, naclSecretKey)
	mac.Write(reason)
	symmetricKey := mac.Sum(nil)
```

  * Add a new type to `libkb/constants.go`:

```
```

  * Add a new constant to `libkb/constants.go`:

```
   EncryptionReasonChatLocalStorage EncryptionReason = "Keybase-Chat-Local-Storage-1"
```

  * Then it's safe to use the message ID as a nonce, so there's no extra overhead to storing a random 24-byte nonce.

How does this sound to everyone?

**Comment by oconnor663** [comment]
I'm indifferent between "HMAC the secret key bytes with some constant plaintext" and "NaCl box some zero bytes to yourself with some constant nonce". That said, I'm not a big fan of the "message ID is the encryption nonce" thing. There are just so many ways that could go wrong. (Does any codepath ever store a message _before_ we get the ID from the server? Would the ID be all zeroes in that case? Could any other bug anywhere in the stack lead to a repeating ID value, or to multiply slightly different copies of the same message? I think that nonce reuse destroys authenticity for _all_ messages, not just the ones that happened to reuse the nonce.)

**Comment by maxtaco** [comment]
OK, let's go with @oconnor663's suggestion above.

**Comment by mmaxim** [comment]
Ok this is updated, let me know what you think.

File: go/libkb/naclwrap.go [hunk 1]
```diff
@@ -680,6 +691,19 @@ func (k NaclDHKeyPair) EncryptToString(plaintext []byte, sender GenericKey) (str
 	return PacketArmoredEncode(info)
 }
 
+func (k NaclDHKeyPair) SecretSymmetricKey(reason EncryptionReason) ([]byte, error) {
+
+	if !k.CanDecrypt() {
+		return nil, KeyCannotDecryptError{}
+	}
```
Comment (maxtaco) [inline go/libkb/naclwrap.go]: To sanity check, should we also require that `len(reason) > 8` or so?  Just to make sure we're not passing something silly in here.

File: go/chat/storage.go [hunk 2]
```diff
@@ -279,22 +332,114 @@ func (s *Storage) updateSupersededBy(bi blockIndex, msgs []chat1.MessageFromServ
...
+
+	return res, nil
+}
+
+func (s *Storage) decryptMessages(ctx context.Context, uid gregor1.UID,
+	msgs []boxedLocalMessage) ([]chat1.MessageFromServerOrError, libkb.ChatStorageError) {
+
+	s.debug("decrypting %d messages", len(msgs))
+
+	// Get decryption key
```
Comment (maxtaco) [inline go/chat/storage.go]: assert that we're getting version=1 back out here since we don't know how to decrypt 2.

Comment (mmaxim) [inline go/chat/storage.go]: Done

**Comment by maxtaco** [comment]
👍

**Pull request merged**